### ISSUE TITLE
PROPOSAL: `download-product` can download from external blobstores (ie s3) with same configuration file

### ISSUE THREAD

**Issue by jtarchie** [pivotal-cf/om#319]
# Problem

For Ops Manager users in air-gapped environments, downloading products via the concourse `s3-resource` can be difficult. For example, the regex to match the `*.pivotal` file can match newer versions than what they'd like to install on the current OpsManager.

The `download-product` config file is an external configuration file, which allows explicit configuration of product, a version or version range, and the latest stemcell to be downloaded. This works well when the environment can connect to Pivnet.

We'd like to use the same config file to support downloading assets from an s3 blobstore.

# Proposed Change

Pivnet provides an API to download a product and associated stemcell. This same API cannot be replicated in s3. When the product files (as-is) are uploaded to s3, there is no metadata associating with the product slug and exact Pivnet version number.

When `download-product` is used to download a file from Pivnet, the outputted file would be prepended with the Pivnet product version specific information.

For example, with the product configuration:

```yaml
---
pivnet-api-token: some-token
pivnet-file-glob: "*vsphere*ova"
pivnet-product-slug: ops-manager
product-version: 2.4.3
```

The outputted `*.ova` file would like `ops-manager-2.4.3-pcf-vsphere-2.4-build.145.ova` instead of `pcf-vsphere-2.4-build.145.ova`.

This file would be stored in s3, via a concourse `put` or `s3` CLI.

Within the air-gapped environment, additional configuration is required to connect to an s3 bucket.

For example,

```yaml
pivnet-api-token: some-token
pivnet-file-glob: "*vsphere*ova"
pivnet-product-slug: ops-manager
product-version: 2.4.3
s3-bucket: my-bucket
s3-access-key-id: some-token
s3-secret-access-key: some-token
s3-region-name: region
s3-endpoint: https://example.com
s3-disable-ssl: false
s3-skip-ssl-verification: false
s3-use-v2-signing: true
```

When invoking `download-product` the flag for `--blobstore s3` will be provided to force usage of s3 instead of pivnet.

Then `download-product` would look for all files prepended with `ops-manager-2.4.3` and then use the `pivnet-file-glob` to narrow down selection.

# Notes

* The outputted file would be a different format from current `download-product`. This is a breaking change for anyone doing a `put` in a pipeline. We could only do this action we know this will also be used for an air-gapped environment -- perhaps if `s3` and `pivnet` credentials are detected.
* Stemcell support is coming. _tl;dr_: we'd look at the tile metadata, and search s3 for the latest tile. It doesn't require the same prepending.
* We'd follow a similar pattern for Azure and GCS. We are just doing s3 first.

**Comment by kcboyle** [comment]
why `--air-gapped` instead of something like `--blobstore`?
"Air Gapped" has certain meanings for some customers, which differ based on who is using the term. It can also have certain security connotations that may or may not be correct. 

Thinking ahead, is the pattern for azure/gcs going to be discerned by different config key names, or are we thinking different flags?

**Comment by jtarchie** [comment]
This is feature is being worked on. We have been able to achieve everything from above. 

We've also added support for v2 signing, so that it may with older s3-compatible blob stores.

**Comment by kcboyle** [comment]
This update has been updated and released in https://github.com/pivotal-cf/om/releases/tag/0.52.0

**Comment by jtarchie** [comment]
We're making substantial changes to this from its initial release, so we're re-opening this issue. I've updated the original description to reflect our working notes.

**Comment by kcboyle** [comment]
Download-product has been updated and released!:https://github.com/pivotal-cf/om/releases/tag/0.54.0
The biggest changes are around the prefixing that we were using. Instead of `{slug-version}-restoffilename.tgz`, where if the file somewhat matched we didn't prefix, we are _always_ prefixing with `[slug, version]originalfilename.tgz`. This was done to make the prefix more obvious. This prefixing method also allowed for this command to be more stable and fixed a lot of bugs from 0.52.0.
